File: clients/cli-go/internal/api/progress.go
package api

import (
	"context"
	"net/url"
	"strconv"
	"time"
)

type ProgressQuery struct {
	LearningSpaceID string `json:"learning_space_id,omitempty"`
	Global          bool   `json:"global,omitempty"`
	GoalID          string `json:"goal_id,omitempty"`
	Status          string `json:"status,omitempty"`
	Order           string `json:"order,omitempty"`
	Cursor          string `json:"cursor,omitempty"`
	Limit           int    `json:"limit,omitempty"`
}
type RouteProgress struct {
	Route               RouteRevision `json:"route"`
	CompletedSteps      []string      `json:"completed_steps"`
	Numerator           int           `json:"numerator"`
	Denominator         int           `json:"denominator"`
	Percent             *float64      `json:"percent"`
	Basis               string        `json:"basis"`
	CurrentGoalRevision bool          `json:"current_goal_revision"`
}
type PendingAssessment struct {
	AssessmentID   string   `json:"assessment_id"`
	NodeRevisionID string   `json:"node_revision_id"`
	Reasons        []string `json:"reasons"`
}
type GoalProgress struct {
	LearningSpaceID  string              `json:"learning_space_id"`
	SpaceName        string              `json:"space_name"`
	EvidenceBasis    string              `json:"evidence_basis"`
	EvidenceSources  []EvidenceSource    `json:"evidence_sources"`
	Goal             GoalRevision        `json:"goal"`
	Routes           []RouteProgress     `json:"routes"`
	Nodes            []NodeReduction     `json:"nodes"`
	Sessions         []SessionSummary    `json:"sessions"`
	Recent           []TimelineItem      `json:"recent_activity"`
	RecentHasMore    bool                `json:"recent_has_more"`
	Reviews          []ReviewSchedule    `json:"reviews"`
	Pending          []PendingAssessment `json:"pending_assessments"`
	EvidenceCount    int                 `json:"evidence_count"`
	EstimatedSeconds int64               `json:"estimated_active_seconds"`
	Estimated        bool                `json:"estimated"`
}

type EvidenceSource struct {
	EvidenceID          string `json:"evidence_id"`
	GoalRevisionID      string `json:"goal_revision_id"`
	KnowledgeRevisionID string `json:"knowledge_revision_id"`
	NodeRevisionID      string `json:"node_revision_id"`
	CurrentGoalRevision bool   `json:"current_goal_revision"`
}
type ProgressPage struct {
	Metadata   ProjectionMetadata `json:"metadata"`
	UpdatedAt  time.Time          `json:"updated_at"`
	HighWater  int64              `json:"committed_event_high_water"`
	Items      []GoalProgress     `json:"items"`
	Total      int                `json:"total"`
	NextCursor string             `json:"next_cursor,omitempty"`
}

func progressValues(q ProgressQuery) url.Values {
	return url.Values{"global": {strconv.FormatBool(q.Global)}, "goal_id": {q.GoalID}, "status": {q.Status}, "order": {q.Order}, "cursor": {q.Cursor}, "limit": {strconv.Itoa(q.Limit)}}
}
func (c *Client) Progress(ctx context.Context, q ProgressQuery) (ProgressPage, error) {
	if q.LearningSpaceID != "" {
		c = c.WithLearningSpace(q.LearningSpaceID)
	}
	var result ProgressPage
	err := c.doJSON(ctx, "GET", "/v1/learning/progress?"+progressValues(q).Encode(), true, nil, map[int]bool{200: true}, true, &result)
	if err == nil && (result.Items == nil || result.Total < len(result.Items) || result.Metadata.Generation == "") {
		err = &ProtocolError{Category: "invalid_progress_page"}
	}
	return result, err
}
func (c *Client) ScopedReviews(ctx context.Context, q ProgressQuery, due *time.Time) (ReviewsPage, error) {
	if q.LearningSpaceID != "" {
		c = c.WithLearningSpace(q.LearningSpaceID)
	}
	var result ReviewsPage
	values := progressValues(q)
	values.Del("order")
	if due != nil {
		values.Set("due_before", due.UTC().Format(time.RFC3339Nano))
	}
	err := c.doJSON(ctx, "GET", "/v1/learning/reviews?"+values.Encode(), true, nil, map[int]bool{200: true}, true, &result)
	if err == nil && (result.Items == nil || result.Total < len(result.Items)) {
		err = &ProtocolError{Category: "invalid_reviews_page"}
	}
	return result, err
}

File: clients/web/src/api/progress.ts
import { z } from 'zod'
import { goalSchema } from './runtime'

const id = z.uuid()
const count = z.number().int().nonnegative()
export const progressSearch = z.object({
  space: id.optional(),
  goal: id.optional(),
  status: z.enum(['active', 'draft', 'paused', 'completed', 'archived', 'all']).catch('active'),
  order: z.enum(['priority', 'recent']).catch('priority'),
  due: z.iso.datetime({ offset: true }).optional(),
  view: z.enum(['goals', 'reviews']).catch('goals'),
})
export type ProgressSearch = z.infer<typeof progressSearch>
export const projectionSchema = z.object({
  generation: id,
  as_of_event_seq: count,
  rebuilding: z.boolean(),
  degraded: z.boolean(),
  incomplete: z.boolean(),
  reason_codes: z.array(z.string()),
})
export const reviewSchema = z.object({
  task_id: id,
  learning_space_id: id,
  goal_id: id,
  goal_revision_id: id,
  knowledge_revision_id: id.optional(),
  route_revision_id: id.optional(),
  session_id: id.optional(),
  carrier_session_id: id.optional(),
  evidence_id: id,
  attempt_id: id.optional(),
  node_revision_id: id,
  goal_name: z.string().optional(),
  space_name: z.string().optional(),
  due_at: z.string(),
  startable: z.boolean().optional(),
  unavailable_reason: z.string().optional(),
})
export type Review = z.infer<typeof reviewSchema>
export const goalProgressSchema = z.object({
  learning_space_id: id,
  space_name: z.string(),
  goal: goalSchema,
  routes: z.array(z.object({
    route: z.object({ route_revision_id: id, route_id: id, revision: count, created_at: z.string(), steps: z.array(z.object({ route_step_id: id, node_revision_id: id, teaching_intent: z.string(), completion_condition: z.string() })) }),
    numerator: count,
    denominator: count,
    percent: z.number().nullable(),
    completed_steps: z.array(id),
    previous_revision_id: id.optional(),
    added_steps: z.array(id),
    removed_steps: z.array(id),
  })),
  sessions: z.array(z.object({ session_id: id, learning_space_id: id, goal_id: id, goal_revision_id: id, name: z.string(), state: z.string(), position: z.string(), resumable: z.boolean(), activity_id: id.optional(), route_revision_id: id.optional(), route_step_id: id.optional() })),
  nodes: z.array(z.object({ mastery: z.object({ node_revision_id: id, state: z.string(), valid_evidence_count: count, pending_assessments: count, uncertainty_reasons: z.array(z.string()) }) })),
  reviews: z.array(reviewSchema),
  pending_assessments: z.array(z.object({ assessment_id: id, node_revision_id: id, reasons: z.array(z.string()) })),
  recent_activity: z.array(z.object({ event_id: id, event_type: z.string(), aggregate_id: id, parent_session_id: id.optional(), received_at: z.string() })),
  recent_has_more: z.boolean(),
  evidence_count: count,
  evidence_sources: z.array(z.object({ evidence_id: id, attempt_id: id.optional(), goal_revision_id: id, knowledge_revision_id: id, current_goal_revision: z.boolean() })),
  estimated_active_seconds: count,
  estimated: z.literal(true),
})
export type GoalProgress = z.infer<typeof goalProgressSchema>
export const progressPage = z.object({ data_cleared: z.boolean().default(false), metadata: projectionSchema, updated_at: z.string(), committed_event_high_water: count, items: z.array(goalProgressSchema), total: count, next_cursor: z.string().optional() })
export const reviewsPage = z.object({ data_cleared: z.boolean().default(false), metadata: projectionSchema, updated_at: z.string(), due_before: z.string(), items: z.array(reviewSchema), total: count, next_cursor: z.string().optional() })
export const reviewReasons: Record<string, string> = {
  original_session_not_at_review_node: '原会话已完成、正在作答或已离开原节点。可查看原记录，或明确创建复习承载。',
  goal_or_space_not_active: '目标已暂停、完成或归档，或学习区已归档。任务与到期日期仍保留。',
  not_due: '尚未到期，保留原调度日期。',
}
export function progressQuery(search: ProgressSearch, cursor: string, limit: number) {
  return { global: !search.space, goal_id: search.goal, status: search.status, cursor, limit }
}
